cmd/gitserver/internal/git/gitcli/odb.go

```diff
@@ -276,3 +284,285 @@ func (g *gitCLIBackend) FirstEverCommit(ctx context.Context) (api.CommitID, erro
 }
 
 const revListUsageString = `usage: git rev-list [<options>] <commit>... [--] [<path>...]`
+
+func (g *gitCLIBackend) Stat(ctx context.Context, commit api.CommitID, path string) (_ fs.FileInfo, err error) {
+	if err := checkSpecArgSafety(string(commit)); err != nil {
+		return nil, err
+	}
+
+	path = filepath.Clean(rel(path))
+
+	// Special case root, which is not returned by `git ls-tree`.
+	if path == "" || path == "." {
+		rev, err := g.revParse(ctx, string(commit)+"^{tree}")
+		if err != nil {
+			if errors.HasType(err, &gitdomain.RevisionNotFoundError{}) {
+				return nil, &os.PathError{Op: "ls-tree", Path: path, Err: os.ErrNotExist}
+			}
+			return nil, err
+		}
+		oid, err := decodeOID(rev)
+		if err != nil {
+			return nil, err
+		}
+		return &fileutil.FileInfo{Mode_: os.ModeDir, Sys_: objectInfo(oid)}, nil
+	}
+
+	it, err := g.lsTree(ctx, commit, path, false)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		closeErr := it.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	fi, err := it.Next()
+	if err != nil {
+		if err == io.EOF {
+			return nil, &os.PathError{Op: "ls-tree", Path: path, Err: os.ErrNotExist}
+		}
+		return nil, err
+	}
+
+	return fi, nil
+}
+
+func (g *gitCLIBackend) ReadDir(ctx context.Context, commit api.CommitID, path string, recursive bool) (git.ReadDirIterator, error) {
+	if err := checkSpecArgSafety(string(commit)); err != nil {
+		return nil, err
+	}
+
+	if path != "" {
+		// Trailing slash is necessary to ls-tree under the dir (not just
+		// to list the dir's tree entry in its parent dir).
+		path = filepath.Clean(rel(path)) + "/"
```

Comment:
Nit, why not use the os path separator?

```suggestion
		path = filepath.Clean(rel(path)) + string(os.PathSeparator)
```

Reply:
On windows, this will be `\` which git might not interpret correctly I believe 👀 (although, I straight up copied that from the existing implementation in the client)